src/unittest/test_threading.cpp

```diff
@@ -0,0 +1,86 @@
+/*
+Minetest
+Copyright (C) 2013 celeron55, Perttu Ahola <celeron55@gmail.com>
+
+This program is free software; you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation; either version 2.1 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License along
+with this program; if not, write to the Free Software Foundation, Inc.,
+51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+*/
+
+#include "test.h"
+
+#include "threading/atomic.h"
+#include "threading/semaphore.h"
+#include "threading/thread.h"
+
+
+class TestThreading : public TestBase {
+public:
+	TestThreading() { TestManager::registerTestModule(this); }
+	const char *getName() { return "TestThreading"; }
+	void runTests(IGameDef *);
+	void testAtomicSemaphoreThread();
+};
+
+static TestThreading g_test_instance;
+
+void TestThreading::runTests(IGameDef *gamedef)
+{
+	TEST(testAtomicSemaphoreThread);
+}
+
+
+class AtomicTestThread : public Thread
+{
+public:
+	AtomicTestThread(Atomic<u32> &v, Semaphore &trigger) :
+		Thread("AtomicTest"),
+		val(v),
+		trigger(trigger)
+	{}
+private:
+	void *run()
+	{
+		trigger.wait();
+		for (u32 i = 0; i < 0x10000; ++i)
+			++val;
```

Comment:
This is okay since it's really a matter of preference, but I just want to let you know that pre-increment is only /required/ for non-primitive types (i.e. iterators).